File: worker/src/handlers.js
// Обработка входящих Telegram-апдейтов (команды и нажатия кнопок).
// Доступ только у владельца (env.OWNER_ID).

import {
  TRACK_ORDER, TRACK_TITLES, TRACK_CSCA, DEFAULT_TZ,
} from "./config.js";
import { CSCA_RESOURCES } from "./content.js";
import * as db from "./db.js";
import * as sched from "./scheduling.js";
import * as rem from "./reminders.js";
import * as stats from "./stats.js";
import * as tg from "./telegram.js";

function authorized(env, userId) {
  const owner = Number(env.OWNER_ID || 0);
  return !owner || userId === owner;
}

// Проверка корректности пояса через попытку форматирования.
function validTz(tz) {
  try {
    new Intl.DateTimeFormat("en-US", { timeZone: tz });
    return true;
  } catch {
    return false;
  }
}

export async function handleUpdate(env, update) {
  if (update.message) return handleMessage(env, update.message);
  if (update.callback_query) return handleCallback(env, update.callback_query);
}

// --- команды ---------------------------------------------------------------
async function handleMessage(env, msg) {
  const from = msg.from;
  const text = (msg.text || "").trim();
  if (!from || !text.startsWith("/")) return;
  if (!authorized(env, from.id)) {
    await tg.sendMessage(env, msg.chat.id, "Это личный бот. Доступа нет.");
    return;
  }

  const [cmdRaw, ...args] = text.split(/\s+/);
  const cmd = cmdRaw.split("@")[0].toLowerCase();
  const uid = from.id;
  const chatId = msg.chat.id;

  switch (cmd) {
    case "/start": return cmdStart(env, uid, chatId);
    case "/help": return cmdHelp(env, chatId);
    case "/status": return cmdStatus(env, uid, chatId);
    case "/tracks": return cmdTracks(env, uid, chatId);
    case "/pause": return cmdPause(env, uid, chatId, args);
    case "/resume": return cmdResume(env, uid, chatId);
    case "/tz": return cmdTz(env, uid, chatId, args);
    case "/report": return cmdReport(env, uid, chatId);
    case "/csca": return cmdCsca(env, chatId);
    default: return tg.sendMessage(env, chatId, "Не знаю такую команду. /help");
  }
}

async function initFires(env, uid, tz) {
  for (const track of await db.getTracks(env, uid)) {
    if (track.next_fire_at == null && track.enabled) {
      await db.setNextFire(env, uid, track.track_id, sched.computeFirstFire(track, tz));
    }
  }
}

async function cmdStart(env, uid, chatId) {
  const isNew = await db.ensureUser(env, uid, chatId);
  const prof = await db.getProfile(env, uid);
  await initFires(env, uid, prof.tz);
  const hello = isNew ? "Привет! " : "С возвращением! ";
  const text = `${hello}Я держу твой ритм по трём трекам:\n\n` +
    "🎧 *6 Minute English* — каждый день, 10 мин\n" +
    "📐 *CSCA Math* — раз в 2 дня, ~40 мин\n" +
    "🧮 *Пробник ЕГЭ* — раз в неделю, ~2 ч\n\n" +
    "Воскресенье — выходной, напоминаний нет.\n\n" +
    "Команды: /status /tracks /pause /report /csca /help";
  await tg.sendMessage(env, chatId, text);
}

function cmdHelp(env, chatId) {
  return tg.sendMessage(env, chatId,
    "*Команды*\n" +
    "/status — план на сегодня и стрик\n" +
    "/tracks — треки: вкл/выкл и расписание\n" +
    "/pause N — пауза на N дней (по умолчанию 1)\n" +
    "/resume — снять паузу\n" +
    "/tz Asia/Yekaterinburg — часовой пояс\n" +
    "/report — недельный отчёт сейчас\n" +
    "/csca — ресурсы по математике\n\n" +
    "В каждом напоминании: [🔗 ссылка] [✅ Готово] [🕑 Позже] [⏭ Пропустить].");
}

async function cmdStatus(env, uid, chatId) {
  const prof = await db.getProfile(env, uid);
  if (!prof) return tg.sendMessage(env, chatId, "Напиши /start сначала.");
  const tz = prof.tz;
  const now = Date.now();
  const info = sched.localInfo(now, tz);
  const streak = await stats.computeStreak(env, uid, tz);

  const lines = [`📍 *Сегодня* (${info.dateStr}, ${tz})`, ""];
  if (!prof.active) {
    const when = prof.paused_until ? sched.localInfo(prof.paused_until, tz).dateStr : "?";
    lines.push(`⏸ На паузе до ${when}. /resume чтобы включить.`);
    return tg.sendMessage(env, chatId, lines.join("\n"));
  }
  if (sched.isQuietDay(now, tz)) lines.push("🌙 Воскресенье — выходной. Напоминаний не будет.");

  const tracks = await db.getTracks(env, uid);
  const byId = Object.fromEntries(tracks.map((t) => [t.track_id, t]));
  for (const tid of sched.orderTracks(tracks.map((t) => t.track_id))) {
    const tr = byId[tid];
    if (!tr.enabled) continue;
    const title = TRACK_TITLES[tid];
    if (tr.next_fire_at) {
      const i = sched.localInfo(tr.next_fire_at, tz);
      lines.push(`• ${title}: → ${i.dateStr} ${String(i.hh).padStart(2, "0")}:${String(i.mm).padStart(2, "0")}`);
    } else {
      lines.push(`• ${title}: не запланирован`);
    }
  }
  lines.push("");
  lines.push(`${streak ? "🔥" : "•"} Стрик: *${streak}* дн.`);
  await tg.sendMessage(env, chatId, lines.join("\n"));
}

function cadenceHuman(track) {
  let base;
  if (track.cadence === "daily") base = "каждый день";
  else if (track.cadence === "weekly") {
    const wd = ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"][track.weekday];
    base = `раз в неделю (${wd})`;
  } else base = `раз в ${track.n_days} дн.`;
  return `${base}, ${String(track.hour).padStart(2, "0")}:${String(track.minute).padStart(2, "0")}`;
}

async function tracksKeyboard(env, uid) {
  const rows = [];
  for (const tid of TRACK_ORDER) {
    const tr = await db.getTrack(env, uid, tid);
    if (!tr) continue;
    const mark = tr.enabled ? "🟢" : "⚪️";
    rows.push([{ text: `${mark} ${TRACK_TITLES[tid]} — ${cadenceHuman(tr)}`, callback_data: `noop:${tid}` }]);
    rows.push([
      { text: tr.enabled ? "Выключить" : "Включить", callback_data: `toggle:${tid}` },
      { text: "−1 ч", callback_data: `hour:${tid}:-1` },
      { text: "+1 ч", callback_data: `hour:${tid}:1` },
    ]);
  }
  return { inline_keyboard: rows };
}

async function cmdTracks(env, uid, chatId) {
  if (!(await db.getProfile(env, uid))) return tg.sendMessage(env, chatId, "Напиши /start сначала.");
  await tg.sendMessage(env, chatId, "*Треки*", { replyMarkup: await tracksKeyboard(env, uid) });
}

async function cmdPause(env, uid, chatId, args) {
  let days = 1;
  if (args[0] && /^\d+$/.test(args[0])) days = Math.max(1, parseInt(args[0], 10));
  const until = Date.now() + days * 86400000;
  await db.setActive(env, uid, 0, until);
  const prof = await db.getProfile(env, uid);
  const i = sched.localInfo(until, prof.tz);
  await tg.sendMessage(env, chatId,
    `⏸ Пауза на ${days} дн. — до ${i.dateStr} ${String(i.hh).padStart(2, "0")}:${String(i.mm).padStart(2, "0")}. Включу сам, или /resume раньше.`);
}

async function cmdResume(env, uid, chatId) {
  await db.setActive(env, uid, 1, null);
  const prof = await db.getProfile(env, uid);
  for (const track of await db.getTracks(env, uid)) {
    if (track.enabled) await db.setNextFire(env, uid, track.track_id, sched.computeFirstFire(track, prof.tz));
  }
  await tg.sendMessage(env, chatId, "▶️ Погнали. Напоминания снова включены.");
}

async function cmdTz(env, uid, chatId, args) {
  if (!args[0]) {
    const prof = await db.getProfile(env, uid);
    return tg.sendMessage(env, chatId, `Часовой пояс: ${prof ? prof.tz : DEFAULT_TZ}\nСменить: /tz Asia/Yekaterinburg`);
  }
  const tz = args[0];
  if (!validTz(tz)) return tg.sendMessage(env, chatId, "Не знаю такой пояс. Пример: /tz Asia/Yekaterinburg");
  await db.setTz(env, uid, tz);
  for (const track of await db.getTracks(env, uid)) {
    if (track.enabled) await db.setNextFire(env, uid, track.track_id, sched.computeFirstFire(track, tz));
  }
  await tg.sendMessage(env, chatId, `Ок, часовой пояс: ${tz}. Расписание пересчитал.`);
}

async function cmdReport(env, uid, chatId) {
  const prof = await db.getProfile(env, uid);
  if (!prof) return tg.sendMessage(env, chatId, "Напиши /start сначала.");
  await tg.sendMessage(env, chatId, await stats.weeklyReport(env, uid, prof.tz));
}

function cmdCsca(env, chatId) {
  const lines = ["📐 *CSCA Math — ресурсы*", ""];
  for (const [name, url, note] of CSCA_RESOURCES) lines.push(`• [${name}](${url}) — ${note}`);
  lines.push("");
  lines.push("_Правило: сначала прогоняй пробники и лови слабые темы, потом добивай именно их._");
  return tg.sendMessage(env, chatId, lines.join("\n"), { disablePreview: true });
}

// --- callbacks -------------------------------------------------------------
async function handleCallback(env, q) {
  const from = q.from;
  if (!authorized(env, from.id)) {
    await tg.answerCallbackQuery(env, q.id, "Нет доступа");
    return;
  }
  const uid = from.id;
  const chatId = q.message.chat.id;
  const messageId = q.message.message_id;
  const baseText = q.message.text || "";
  const parts = (q.data || "").split(":");
  const action = parts[0];

  if (action === "noop") return tg.answerCallbackQuery(env, q.id);

  if (action === "done" || action === "skip") {
    const hid = Number(parts[1]);
    const h = await db.getHistory(env, hid);
    if (!h) return tg.answerCallbackQuery(env, q.id, "Уже неактуально");
    if (action === "done") {
      await db.setHistoryStatus(env, hid, "done");
      const prof = await db.getProfile(env, uid);
      const streak = await stats.computeStreak(env, uid, prof.tz);
      await tg.answerCallbackQuery(env, q.id, "Готово 🎯");
      await finalize(env, chatId, messageId, baseText, `\n\n✅ Готово. 🔥 Стрик: ${streak} дн.`);
    } else {
      await db.setHistoryStatus(env, hid, "skipped");
      await tg.answerCallbackQuery(env, q.id, "Пропущено");
      await finalize(env, chatId, messageId, baseText, "\n\n⏭ Пропущено. Ничего, завтра новый заход.");
    }
    return;
  }

  if (action === "later") {
    const hid = Number(parts[1]);
    await tg.answerCallbackQuery(env, q.id);
    return tg.editMessageReplyMarkup(env, chatId, messageId, rem.laterKeyboard(hid));
  }

  if (action === "back") {
    const hid = Number(parts[1]);
    const h = await db.getHistory(env, hid);
    await tg.answerCallbackQuery(env, q.id);
    if (h) return tg.editMessageReplyMarkup(env, chatId, messageId, rem.actionKeyboard(hid, h.content_ref));
    return;
  }

  if (action === "post2h" || action === "posttom") {
    const hid = Number(parts[1]);
    const h = await db.getHistory(env, hid);
    if (!h) return tg.answerCallbackQuery(env, q.id, "Уже неактуально");
    await db.setHistoryStatus(env, hid, "postponed");
    const prof = await db.getProfile(env, uid);
    const kind = action === "post2h" ? "2h" : "tomorrow";
    const newMs = sched.postpone(kind, prof.tz, Date.now());
    await db.setNextFire(env, uid, h.track_id, newMs);
    const i = sched.localInfo(newMs, prof.tz);
    await tg.answerCallbackQuery(env, q.id, "Перенёс");
    await finalize(env, chatId, messageId, baseText,
      `\n\n🕑 Перенесено на ${i.dateStr} ${String(i.hh).padStart(2, "0")}:${String(i.mm).padStart(2, "0")}.`);
    return;
  }

  if (action === "why") {
    const hid = Number(parts[1]);
    const reason = parts[2];
    const h = await db.getHistory(env, hid);
    if (!h) return tg.answerCallbackQuery(env, q.id, "Уже неактуально");
    await db.setHistoryStatus(env, hid, "skipped", reason);
    if (reason === "bored") {
      await db.bumpContentIndex(env, uid, h.track_id);
      const prof = await db.getProfile(env, uid);
      await db.setNextFire(env, uid, h.track_id, Date.now() + 2 * 86400000);
    }
    await tg.answerCallbackQuery(env, q.id, "Принял");
    await finalize(env, chatId, messageId, baseText, "\n\n" + rem.whyResponse(h.track_id, reason));
    return;
  }

  if (action === "toggle") {
    const tid = parts[1];
    const tr = await db.getTrack(env, uid, tid);
    const newState = tr.enabled ? 0 : 1;
    await db.updateTrack(env, uid, tid, { enabled: newState });
    const prof = await db.getProfile(env, uid);
    if (newState) {
      await db.setNextFire(env, uid, tid, sched.computeFirstFire(await db.getTrack(env, uid, tid), prof.tz));
    } else {
      await db.setNextFire(env, uid, tid, null);
    }
    await tg.answerCallbackQuery(env, q.id, "Ок");
    return tg.editMessageReplyMarkup(env, chatId, messageId, await tracksKeyboard(env, uid));
  }

  if (action === "hour") {
    const tid = parts[1];
    const delta = parseInt(parts[2], 10);
    const tr = await db.getTrack(env, uid, tid);
    const newHour = ((tr.hour + delta) % 24 + 24) % 24;
    await db.updateTrack(env, uid, tid, { hour: newHour });
    const prof = await db.getProfile(env, uid);
    if (tr.enabled) await db.setNextFire(env, uid, tid, sched.computeFirstFire(await db.getTrack(env, uid, tid), prof.tz));
    await tg.answerCallbackQuery(env, q.id, `${String(newHour).padStart(2, "0")}:00`);
    return tg.editMessageReplyMarkup(env, chatId, messageId, await tracksKeyboard(env, uid));
  }

  if (action === "adapt") {
    if (parts[1] === "none") {
      await tg.answerCallbackQuery(env, q.id, "Оставил как есть");
      return finalize(env, chatId, messageId, baseText, "\n\nОк, расписание без изменений.");
    }
    const tid = parts[1];
    const newN = parseInt(parts[2], 10);
    await db.updateTrack(env, uid, tid, { n_days: newN });
    const prof = await db.getProfile(env, uid);
    await db.setNextFire(env, uid, tid, sched.computeFirstFire(await db.getTrack(env, uid, tid), prof.tz));
    await tg.answerCallbackQuery(env, q.id, "Сделал реже");
    return finalize(env, chatId, messageId, baseText, `\n\n✅ ${TRACK_TITLES[tid]}: теперь раз в ${newN} дн.`);
  }

  await tg.answerCallbackQuery(env, q.id);
}

async function finalize(env, chatId, messageId, baseText, suffix) {
  const r = await tg.editMessageText(env, chatId, messageId, baseText + suffix);
  if (!r.ok) await tg.editMessageReplyMarkup(env, chatId, messageId, { inline_keyboard: [] });
}

File: handlers.py
"""Обработчики команд и нажатий кнопок.

Доступ только у владельца (OWNER_ID). Команды:
  /start   — регистрация и приветствие
  /status  — план на сегодня, треки, стрик
  /tracks  — список треков, включить/выключить, поменять время
  /pause N — пауза на N дней (кнопка «на паузу»)
  /resume  — снять паузу
  /tz Zone — сменить часовой пояс
  /report  — недельный отчёт по запросу
  /csca    — ресурсы CSCA
  /help
"""

from datetime import timedelta
from zoneinfo import ZoneInfo, available_timezones

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.constants import ParseMode
from telegram.ext import ContextTypes

import config
import content
import reminders
import scheduling
import stats
from db import from_iso, to_iso, utcnow


def _store(context):
    return context.application.bot_data["store"]


def _authorized(update: Update) -> bool:
    if not config.OWNER_ID:
        return True  # OWNER_ID не задан — не ограничиваем (dev-режим)
    user = update.effective_user
    return user and user.id == config.OWNER_ID


async def _guard(update: Update) -> bool:
    if not _authorized(update):
        if update.message:
            await update.message.reply_text("Это личный бот. Доступа нет.")
        elif update.callback_query:
            await update.callback_query.answer("Нет доступа", show_alert=True)
        return False
    return True


# --- команды ---------------------------------------------------------------
async def cmd_start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    store = _store(context)
    uid = update.effective_user.id
    is_new = store.ensure_user(uid, update.effective_chat.id)
    # Догоняем треки до актуальных дефолтов (новый Duolingo, новые кадентности).
    store.sync_tracks(uid)

    # Инициализируем next_fire для треков, если ещё не заданы.
    prof = store.get_profile(uid)
    for track in store.get_tracks(uid):
        if track["next_fire_at"] is None and track["enabled"]:
            first = scheduling.compute_first_fire(track, prof["tz"])
            store.set_next_fire(uid, track["track_id"], first)

    hello = "Привет! " if is_new else "С возвращением! "
    text = (
        f"{hello}Я держу твой ритм по четырём трекам:\n\n"
        "🎧 *6 Minute English* — раз в 2 дня, 10 мин\n"
        "🦉 *Duolingo Test* — раз в 2 дня, ~20 мин (для вузов Китая)\n"
        "📐 *CSCA Math* — раз в 3 дня, подробный урок EN→RU + практика\n"
        "🧮 *Пробник ЕГЭ* — раз в неделю, ~2 ч\n\n"
        "Воскресенье — выходной, напоминаний нет.\n\n"
        "Команды: /status /tracks /pause /report /csca /duo /help"
    )
    await update.message.reply_text(text, parse_mode=ParseMode.MARKDOWN)


async def cmd_help(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    await update.message.reply_text(
        "*Команды*\n"
        "/status — план на сегодня и стрик\n"
        "/tracks — треки: вкл/выкл и расписание\n"
        "/pause N — пауза на N дней (по умолчанию 1)\n"
        "/resume — снять паузу\n"
        "/tz Asia/Yekaterinburg — часовой пояс\n"
        "/report — недельный отчёт сейчас\n"
        "/csca — ресурсы по математике\n"
        "/duo — про Duolingo Test и ресурсы\n\n"
        "В каждом напоминании: [🔗 ссылка] [✅ Готово] [🕑 Позже] [⏭ Пропустить].",
        parse_mode=ParseMode.MARKDOWN,
    )


async def cmd_status(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    store = _store(context)
    uid = update.effective_user.id
    prof = store.get_profile(uid)
    if not prof:
        await update.message.reply_text("Напиши /start сначала.")
        return
    tz = prof["tz"]
    now_local = scheduling.local_now(tz)
    streak = stats.compute_streak(store, uid, tz)

    lines = [f"📍 *Сегодня* ({now_local:%a %d.%m}, {tz})", ""]
    if not prof["active"]:
        pu = from_iso(prof["paused_until"])
        when = pu.astimezone(ZoneInfo(tz)).strftime("%d.%m %H:%M") if pu else "?"
        lines.append(f"⏸ На паузе до {when}. /resume чтобы включить.")
        await update.message.reply_text("\n".join(lines),
                                        parse_mode=ParseMode.MARKDOWN)
        return

    if scheduling.is_quiet_day(now_local):
        lines.append("🌙 Воскресенье — выходной. Напоминаний не будет.")
    for track in scheduling.order_due([t["track_id"] for t in store.get_tracks(uid)]):
        tr = store.get_track(uid, track)
        if not tr["enabled"]:
            continue
        nf = from_iso(tr["next_fire_at"])
        title = config.TRACK_TITLES[track]
        if nf:
            loc = nf.astimezone(ZoneInfo(tz))
            when = loc.strftime("%d.%m %H:%M")
            lines.append(f"• {title}: → {when}")
        else:
            lines.append(f"• {title}: не запланирован")

    lines.append("")
    fire = "🔥" if streak else "•"
    lines.append(f"{fire} Стрик: *{streak}* дн.")
    await update.message.reply_text("\n".join(lines), parse_mode=ParseMode.MARKDOWN)


async def cmd_tracks(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    store = _store(context)
    uid = update.effective_user.id
    if not store.get_profile(uid):
        await update.message.reply_text("Напиши /start сначала.")
        return
    await update.message.reply_text("*Треки*", parse_mode=ParseMode.MARKDOWN,
                                    reply_markup=_tracks_keyboard(store, uid))


def _cadence_human(track):
    if track["cadence"] == "daily":
        base = "каждый день"
    elif track["cadence"] == "weekly":
        wd = ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"][track["weekday"]]
        base = f"раз в неделю ({wd})"
    else:
        base = f"раз в {track['n_days']} дн."
    return f"{base}, {track['hour']:02d}:{track['minute']:02d}"


def _tracks_keyboard(store, uid):
    rows = []
    for tid in config.TRACK_ORDER:
        tr = store.get_track(uid, tid)
        if not tr:
            continue
        mark = "🟢" if tr["enabled"] else "⚪️"
        title = config.TRACK_TITLES[tid]
        rows.append([InlineKeyboardButton(
            f"{mark} {title} — {_cadence_human(tr)}",
            callback_data=f"trk:{tid}")])
        rows.append([
            InlineKeyboardButton(
                "Выключить" if tr["enabled"] else "Включить",
                callback_data=f"toggle:{tid}"),
            InlineKeyboardButton("−1 ч", callback_data=f"hour:{tid}:-1"),
            InlineKeyboardButton("+1 ч", callback_data=f"hour:{tid}:1"),
        ])
    return InlineKeyboardMarkup(rows)


async def cmd_pause(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    store = _store(context)
    uid = update.effective_user.id
    days = 1
    if context.args:
        try:
            days = max(1, int(context.args[0]))
        except ValueError:
            pass
    until = utcnow() + timedelta(days=days)
    store.set_active(uid, 0, until)
    prof = store.get_profile(uid)
    loc = until.astimezone(ZoneInfo(prof["tz"])).strftime("%d.%m %H:%M")
    await update.message.reply_text(
        f"⏸ Пауза на {days} дн. — до {loc}. Включу сам, или /resume раньше."
    )


async def cmd_resume(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    store = _store(context)
    uid = update.effective_user.id
    store.set_active(uid, 1, None)
    # Пересчитываем ближайшие срабатывания от «сейчас».
    prof = store.get_profile(uid)
    for track in store.get_tracks(uid):
        if track["enabled"]:
            store.set_next_fire(uid, track["track_id"],
                                scheduling.compute_first_fire(track, prof["tz"]))
    await update.message.reply_text("▶️ Погнали. Напоминания снова включены.")


async def cmd_tz(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    store = _store(context)
    uid = update.effective_user.id
    if not context.args:
        prof = store.get_profile(uid)
        await update.message.reply_text(
            f"Часовой пояс: {prof['tz']}\nСменить: /tz Asia/Yekaterinburg")
        return
    tz = context.args[0]
    if tz not in available_timezones():
        await update.message.reply_text("Не знаю такой пояс. Пример: /tz Asia/Yekaterinburg")
        return
    store.set_tz(uid, tz)
    for track in store.get_tracks(uid):
        if track["enabled"]:
            store.set_next_fire(uid, track["track_id"],
                                scheduling.compute_first_fire(track, tz))
    await update.message.reply_text(f"Ок, часовой пояс: {tz}. Расписание пересчитал.")


async def cmd_report(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    store = _store(context)
    uid = update.effective_user.id
    prof = store.get_profile(uid)
    if not prof:
        await update.message.reply_text("Напиши /start сначала.")
        return
    text = stats.weekly_report(store, uid, prof["tz"])
    await update.message.reply_text(text, parse_mode=ParseMode.MARKDOWN)


async def cmd_csca(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    lines = ["📐 *CSCA Math — ресурсы*", ""]
    for name, url, note in content.CSCA_RESOURCES:
        lines.append(f"• [{name}]({url}) — {note}")
    lines.append("")
    lines.append("_Правило: сначала прогоняй пробники и лови слабые темы, "
                 "потом добивай именно их._")
    await update.message.reply_text("\n".join(lines), parse_mode=ParseMode.MARKDOWN,
                                    disable_web_page_preview=True)


async def cmd_duo(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    lines = ["🦉 *Duolingo English Test — для поступления в Китай*", ""]
    lines.append("Адаптивный тест ~1 ч, балл 10–160. *Твоя цель — 110–120* "
                 "(этого хватает большинству программ; топовые вузы иногда просят "
                 "120+). Проверь минимум своих вузов заранее.")
    lines.append("")
    for name, url, note in content.DET_RESOURCES:
        lines.append(f"• [{name}]({url}) — {note}")
    lines.append("")
    lines.append("_План: раз в 2 дня бот присылает конкретный тип задания DET "
                 "с подсказкой. Раз в пару недель — полный пробный тест по кнопке._")
    await update.message.reply_text("\n".join(lines), parse_mode=ParseMode.MARKDOWN,
                                    disable_web_page_preview=True)


# --- callbacks -------------------------------------------------------------
async def on_callback(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not await _guard(update):
        return
    q = update.callback_query
    store = _store(context)
    uid = update.effective_user.id
    data = q.data or ""
    parts = data.split(":")
    action = parts[0]

    if action in ("done", "skip"):
        hid = int(parts[1])
        h = store.get_history(hid)
        if not h:
            await q.answer("Уже неактуально")
            return
        if action == "done":
            store.set_history_status(hid, "done")
            streak = stats.compute_streak(store, uid, store.get_profile(uid)["tz"])
            await q.answer("Готово 🎯")
            suffix = f"\n\n✅ Готово. 🔥 Стрик: {streak} дн."
        else:
            store.set_history_status(hid, "skipped")
            await q.answer("Пропущено")
            suffix = "\n\n⏭ Пропущено. Ничего, завтра новый заход."
        await _finalize(q, suffix)
        return

    if action == "later":
        hid = int(parts[1])
        await q.answer()
        await q.edit_message_reply_markup(reply_markup=reminders.later_keyboard(hid))
        return

    if action == "back":
        hid = int(parts[1])
        h = store.get_history(hid)
        if h:
            await q.edit_message_reply_markup(
                reply_markup=reminders.action_keyboard(hid, h["content_ref"]))
        await q.answer()
        return

    if action in ("post2h", "posttom"):
        hid = int(parts[1])
        h = store.get_history(hid)
        if not h:
            await q.answer("Уже неактуально")
            return
        store.set_history_status(hid, "postponed")
        prof = store.get_profile(uid)
        now_local = scheduling.local_now(prof["tz"])
        kind = "2h" if action == "post2h" else "tomorrow"
        new_local = scheduling.postpone(now_local, kind, prof["tz"])
        # Ставим next_fire трека на момент переноса, чтобы напоминание пришло снова.
        store.set_next_fire(uid, h["track_id"], new_local)
        when = new_local.strftime("%d.%m %H:%M")
        await q.answer("Перенёс")
        await _finalize(q, f"\n\n🕑 Перенесено на {when}.")
        return

    if action == "why":
        hid = int(parts[1])
        reason = parts[2]
        h = store.get_history(hid)
        if not h:
            await q.answer("Уже неактуально")
            return
        store.set_history_status(hid, "skipped", feedback=reason)
        # На «надоел формат» — двигаем контент и даём паузу треку на пару дней.
        if reason == "bored":
            store.bump_content_index(uid, h["track_id"])
            tr = store.get_track(uid, h["track_id"])
            new_local = scheduling.local_now(store.get_profile(uid)["tz"]) + timedelta(days=2)
            store.set_next_fire(uid, h["track_id"], new_local)
        await q.answer("Принял")
        await _finalize(q, "\n\n" + reminders.why_response(h["track_id"], reason))
        return

    if action == "toggle":
        tid = parts[1]
        tr = store.get_track(uid, tid)
        new_state = 0 if tr["enabled"] else 1
        store.update_track(uid, tid, enabled=new_state)
        if new_state:
            prof = store.get_profile(uid)
            store.set_next_fire(uid, tid,
                                scheduling.compute_first_fire(
                                    store.get_track(uid, tid), prof["tz"]))
        else:
            store.set_next_fire(uid, tid, None)
        await q.answer("Ок")
        await q.edit_message_reply_markup(reply_markup=_tracks_keyboard(store, uid))
        return

    if action == "hour":
        tid = parts[1]
        delta = int(parts[2])
        tr = store.get_track(uid, tid)
        new_hour = (tr["hour"] + delta) % 24
        store.update_track(uid, tid, hour=new_hour)
        prof = store.get_profile(uid)
        if tr["enabled"]:
            store.set_next_fire(uid, tid,
                                scheduling.compute_first_fire(
                                    store.get_track(uid, tid), prof["tz"]))
        await q.answer(f"{new_hour:02d}:00")
        await q.edit_message_reply_markup(reply_markup=_tracks_keyboard(store, uid))
        return

    if action == "adapt":
        if parts[1] == "none":
            await q.answer("Оставил как есть")
            await _finalize(q, "\n\nОк, расписание без изменений.")
            return
        tid = parts[1]
        new_n = int(parts[2])
        store.update_track(uid, tid, n_days=new_n)
        prof = store.get_profile(uid)
        store.set_next_fire(uid, tid,
                            scheduling.compute_first_fire(
                                store.get_track(uid, tid), prof["tz"]))
        await q.answer("Сделал реже")
        await _finalize(q, f"\n\n✅ {config.TRACK_TITLES[tid]}: теперь раз в {new_n} дн.")
        return

    await q.answer()


async def _finalize(q, suffix: str):
    """Гасим кнопки и дописываем результат к тексту напоминания."""
    try:
        base = q.message.text or ""
        await q.edit_message_text(base + suffix, parse_mode=ParseMode.MARKDOWN)
    except Exception:
        try:
            await q.edit_message_reply_markup(reply_markup=None)
        except Exception:
            pass
